javatests/arcs/core/crdt/CrdtEntityTest.kt

```diff
@@ -513,4 +514,496 @@ class CrdtEntityTest {
+    val entityB = CrdtEntity(VersionMap(), rawEntityB)
+
+    entityA.merge(entityB.data)
+    entityB.merge(entityAData)
+
+    assertThat(entityA.consumerView.singletons).containsExactlyEntriesIn(
+      mapOf(
+        "foo" to Reference("fooRef"),
+        "fooBar" to Reference("fooBarRef")
+      )
+    )
+    assertThat(entityB.consumerView.singletons).containsExactlyEntriesIn(
+      mapOf(
+        "bar" to Reference("barRef"),
+        "fooBar" to Reference("fooBarRef")
+      )
+    )
+  }
+
+  /**
+   * Test merges where a singleton's types do not match.
+   * Note: this is currently producing unexpected behaviour. The test should be updated once
+   * b/177036049 is resolved.
+   */
+  @Test
+  fun crdtEntity_merge_singletonFieldTypeMismatch() {
+    val rawEntityA = RawEntity(
+      id = "an-id",
+      singletons = mapOf("koalas" to Reference("fooRef")),
+      collections = mapOf()
+    )
+    val rawEntityB = RawEntity(
+      id = "an-id",
+      singletons = mapOf(
+        "koalas" to ReferencablePrimitive(Double::class, 1.0)
+      ),
+      collections = mapOf()
+    )
+
+    val entityA = CrdtEntity(VersionMap("me" to 1), rawEntityA)
+    val entityAData = entityA.data
+    val entityB = CrdtEntity(VersionMap("me" to 2), rawEntityB)
+
+    entityA.merge(entityB.data)
+    entityB.merge(entityAData)
+
+    assertThat(entityA.consumerView.singletons["foo"]).isNull()
+    assertThat(entityB.consumerView.singletons["foo"]).isNull()
+  }
+
+  /**
+   * Test that the merge function is commutative.
+   * i.e. A.merge(B).merge(C) = A.merge(C).merge(B)
+   */
+  @Test
+  fun crdtEntity_merge_commutive() {
+    val rawEntityA = RawEntity(
+      id = "an-id",
+      singletons = mapOf("foo" to Reference("fooRef")),
+      collections = mapOf(
+        "bar" to setOf(Reference("barRef1"), Reference("barRef2"))
+      )
+    )
+    val entityA1 = CrdtEntity(VersionMap(), rawEntityA)
+    val entityA2 = CrdtEntity(VersionMap(), rawEntityA)
+    val mergedEntity1 = CrdtEntity(
+      VersionMap(),
+      RawEntity(
+        id = "an-id",
+        singletons = mapOf("foo" to Reference("fooRefMerge1")),
+        collections = mapOf(
+          "bar" to setOf(Reference("barRef1Merge1"), Reference("barRef2Merge1"))
+        )
+      )
+    )
+    val mergedEntity2 = CrdtEntity(
+      VersionMap(),
+      RawEntity(
+        id = "an-id",
+        singletons = mapOf("foo" to Reference("fooRefMerge2")),
+        collections = mapOf(
+          "bar" to setOf(Reference("barRef1Merge2"), Reference("barRef2Merge2"))
+        )
+      )
+    )
+
+    entityA1.merge(mergedEntity1.data)
+    entityA1.merge(mergedEntity2.data)
+
+    entityA2.merge(mergedEntity2.data)
+    entityA2.merge(mergedEntity1.data)
+
+    assertThat(entityA1.data).isEqualTo(entityA2.data)
+    assertThat(entityA1.versionMap).isEqualTo(entityA2.versionMap)
+  }
+
+  /**
+   * Verify that the merge operation is associative.
+   * i.e. A.merge(B) = B.merge(A)
```

Comment:
the first test seemingly violates this :)

Reply:
Fixed :)